### PR TITLE
Add manual application

### PR DESCRIPTION
**Pull request by caelwithcats**
The man pages are stored in /System/Configs/man-pages.json, there is still work to be done in this file(adding more entries)

```
$ man echo
echo -
DESCRIPTION:
prints text to the terminal
USAGE:
echo [STRING]
EXAMPLE:
echo hello world: hello world
```

### CODE REVIEW

**Comment by caelwithcats** [comment]
@cristian-programmer do you mean add comments to my code, make it cleaner? I do not know what "bleed the source code" means

File: coreutils/man.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+           {
+	       JsonValue *manpage = json_object_get(root, argv[i]);
+	       const char* message = "man: It looks like your %s file is corrupted\n";
+
+	       const char* description = "";
+	       const char* usage = "";
+	       const char* example = "";
+	       if (json_object_has(manpage, "description")) { JsonValue *value = json_object_get(manpage, "description"); if(json_is(value, JSON_STRING)) { description = json_string_value(value); } else { printf(message, manpath); return -1; }  } else { printf(message, manpath); return -1; }
+	       if (json_object_has(manpage, "usage")) { JsonValue *value = json_object_get(manpage, "usage"); if(json_is(value, JSON_STRING)) { usage = json_string_value(value); } else { printf(message, manpath); return -1; }  } else { printf(message, manpath); return -1; }
+	       if (json_object_has(manpage, "example")) { JsonValue *value = json_object_get(manpage, "example"); if(json_is(value, JSON_STRING)) { example = json_string_value(value); } else { printf(message, manpath); return -1; }  } else { printf(message, manpath); return -1; }
```
Comment (NicolasVanBossuyt) [inline coreutils/man.cpp:33]: Please wrap your lines around 60~80 column

Comment (caelwithcats) [reply coreutils/man.cpp:33]: ok, that makes sense. I'll fix it

File: sysroot/System/Configs/man-pages.json [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    	"description": "man is short for manual. It provides documentation on commands",
+    	"usage": "man [OPTIONS]",
+    	"example": "man echo uptime"
+    },
+    
+    "sysfetch": {
+    	"description": "grabs system information such as the kernel, version, uptime, shell, etc",
+    	"usage": "sysfetch",
+    	"example": "sysfetch"
+    }
```
Comment (NicolasVanBossuyt) [inline sysroot/System/Configs/man-pages.json:24]: This JSON file is malformed, you have missings `,`

File: coreutils/man.cpp (hunk 1 above)
Comment (cristian-programmer) [reply coreutils/man.cpp:33]: @caelwithcats yes, what monax commented :)

File: sysroot/System/Configs/man-pages.json (hunk 2 above)
Comment (caelwithcats) [reply sysroot/System/Configs/man-pages.json]: thanks for finding that, when I tested it, it ran fine. This is seems to be a bug in the JSON reader I think

Comment (caelwithcats) [reply sysroot/System/Configs/man-pages.json]: oops, I forgot to give a commit name. I wish github allowed you to rename your commits

Comment (NicolasVanBossuyt) [reply sysroot/System/Configs/man-pages.json]: yes, you can if you git push --force

Comment (caelwithcats) [reply sysroot/System/Configs/man-pages.json]: ok, I changed the name by it making a new commit

**Comment by NicolasVanBossuyt** [comment]
@caelwithcats did you see my comments about the location and structure of the json files ?

**Comment by caelwithcats** [comment]
@NicolasVanBossuyt yes, I have and I have been working on it. What do you think:
```c++
#include <libjson/Json.h>
#include <stdio.h>
#include <string.h>
#include <libsystem/io/Stream.h>

int main(int argc, char **argv)
{

	if (argc == 1)
	{
		printf("What manual page do you need? eg. man man\n");
	}
	for (int i = 1; i < argc; i++)
	{
		// This is why c code can be annoying sometimes
		char tmp[strlen("/System/Manuals/") + strlen(argv[i]) + strlen(".json")];
		sprintf(tmp, "/System/Manuals/%s%s", argv[i], ".json");
		const char *manpath = tmp;
		__cleanup(stream_cleanup) Stream *stream = stream_open(manpath, OPEN_READ);
		if (handle_has_error(stream))
		{
			handle_printf_error(stream, "man: No manual entry for \"%s\"", argv[i]);
			return -1;
		}
		JsonValue *root = json_parse_file(manpath);

		if (json_is(root, JSON_OBJECT))
		{
			const char *message = "man: It looks like this man page is corrupted\n";

			const char *description = "";
			const char *usage = "";
			const char *example = "";

			if (json_object_has(root, "description"))
			{
				JsonValue *value = json_object_get(root, "description");
				if (json_is(value, JSON_STRING))
				{
					description = json_string_value(value);
				}
				else
				{
					printf(message);
					return -1;
				}
			}
			else
			{
				printf(message);
				return -1;
			}

			if (json_object_has(root, "usage"))
			{
				JsonValue *value = json_object_get(root, "usage");
				if (json_is(value, JSON_STRING))
				{
					usage = json_string_value(value);
				}
				else
				{
					printf(message);
					return -1;
				}
			}
			else
			{
				printf(message);
				return -1;
			}

			if (json_object_has(root, "example"))
			{
				JsonValue *value = json_object_get(root, "example");
				if (json_is(value, JSON_STRING))
				{
					example = json_string_value(value);
				}
				else
				{
					printf(message);
					return -1;
				}
			}
			else
			{
				printf(message);
				return -1;
			}

			printf("%s -\nDESCRIPTION:\n %s\nUSAGE:\n%s\nEXAMPLE:\n%s\n", argv[i], description, usage, example);
		}
		else
		{
			printf("man: No manual entry for \"%s\"\n", argv[i]);
		}
	}
return 0;
}
```

File: sysroot/System/Configs/man-pages.json (hunk 2 above)
Comment (caelwithcats) [reply sysroot/System/Configs/man-pages.json]: I have followed your suggestion about the structure: https://github.com/skiftOS/skift/pull/84#issuecomment-667859390

**Comment by caelwithcats** [comment]
@NicolasVanBossuyt I am going to move this into a new pull request and close this one.

**Comment by NicolasVanBossuyt** [comment]
Agree this will be cleaner

**Comment by caelwithcats** [comment]
@NicolasVanBossuyt new pull request here: https://github.com/skiftOS/skift/pull/85

**Pull request closed without merging**